### ISSUE TITLE
Avoiding PathTooLongException

### ISSUE THREAD

**Issue by micktion** [dotnet/runtime#61247]
There doesn't seem to be any reliable cross platform method for determining the maximum length of a file on the system using .NET.  

Am I wrong?  Seems like such a basic need, yet I searched and it doesn't seem to be a thing.

For those wanting to avoid exceptions like ...

System.IO.PathTooLongException: The specified path, file name, or both are too long. The fully qualified file name must be less than 260 characters, and the directory name must be less than 248 characters.

Can we have a method of getting the maximum file length permitted by the system...

i.e.
`namespace System.IO
{
    public static class Path
    {
        public static int GetMaxFilePath()
        {
.....`

Than can be called like ...

if (myFilePath.Length <= Path.GetMaxFilePath())
{
    // go go go
}
else
{
   //  don't even try
}

**Comment by TonyValenti** [comment]
I feel like this should live in system.environment.

But yes, I agree. Something like this is needed. I bump in the challenges with long paths all the time.￼

**Comment by Suchiman** [comment]
These limits do not just apply to operating system but also filesystem being used so you'd need something like `Path.IsPossiblePathName(path)`... or you just try to create the path or file name and catch the `PathTooLongException`, or a `Path.GetRemainingNameLength(path, out int remainingDirectoryNameLength, out int remainingFileNameLength)`...

**Comment by micktion** [comment]
Perhaps a method on System.IO.DriveInfo then?

**Comment by adamsitnik** [comment]
How would you implement it on Windows and Unix?

**Comment by Jozkee** [comment]
Are there any solutions available in other languages/frameworks that we can use as a reference?

**Comment by carlossanlop** [comment]
Triage:

- The max supported path length depends largely on the filesystem, and the OS could also have influence. See: https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file#file-and-directory-names
- The concept of a drive is very different between Windows and Unix. The `DriveInfo` API would not be a good option since it doesn't have an ideal behavior in Unix. See https://github.com/dotnet/runtime/issues/26081
- Instead of `System.Environment`, maybe `DirectoryInfo` and `Directory` would be good alternative locations (instead of `DriveInfo`) if we can retrieve filesystem of the specified directory path.
- @micktion can you please provide a proof of concept of what you're proposing? How would we solve your problem with just P/Invokes, both on Windows and on Unix?
- Since you're proposing a new API, here are the guidelines to follow the process: https://github.com/dotnet/runtime/blob/main/docs/project/api-review-process.md

**Comment by Symbai** [comment]
> How would you implement it on Windows and Unix?

On Unix calling [statvfs](https://www.unix.com/man-page/centos/3/STATVFS/) and check `f_namemax` @Jozkee In Python `os.statvfs` which also performs a system call of statvfs

**Comment by micktion** [comment]
TBH I do not know enough about Linux file systems to be able to comment. It definitely looks more challenging to implement for non-windows platforms, considering the numerous file systems available on linux.

**Comment by Symbai** [comment]
You are wrong. You just provide the path as parameter and based on what the file system for this path is, it returns the max file name. As long as this system call exist it works. So for Unix is easy to implement. The only question is does something like this exist on Windows? Even when it does as @carlossanlop said this API request requires more work.
